File: src/slowql/core/scoring.py
# slowql/src/slowql/core/scoring.py
"""
Query complexity scoring logic for SlowQL.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import TYPE_CHECKING

from sqlglot import exp

from slowql.core.models import Severity

if TYPE_CHECKING:
    from slowql.core.models import Issue, Query


class ComplexityScorer:
    """
    Calculates numerical complexity scores for SQL queries.

    The score is based on:
    1. Query structure (joins, subqueries, aggregations)
    2. Detected issues (severity-based weighting)
    """

    def __init__(self, base_score: int = 10):
        self.base_score = base_score

    def calculate_score(self, query: Query, issues: list[Issue]) -> int:
        """Calculate complexity score for a query.

        Args:
            query: The parsed query.
            issues: List of issues found in the query.

        Returns:
            Complexity score (0-100).
        """
        score = self.base_score

        # Add points for query structure
        score += self._calculate_structural_complexity(query)

        # Add points for issues
        score += self._calculate_issue_complexity(issues)

        # Cap at 100
        return min(max(score, 0), 100)

    def _calculate_structural_complexity(self, query: Query) -> int:
        """Calculate complexity based on query structure (joins, subqueries)."""
        score = 0

        # Use AST if available for more accurate analysis
        if query.ast:
            # Joins
            joins = list(query.ast.find_all(exp.Join))
            score += len(joins) * 10

            # Subqueries
            subqueries = list(query.ast.find_all(exp.Subquery))
            score += len(subqueries) * 15

            # Aggregations
            aggregations = list(query.ast.find_all(exp.AggFunc))
            score += len(aggregations) * 5
        else:
            # Fallback to simple regex if AST is not available
            joins = re.findall(r"\bJOIN\b", query.raw, re.IGNORECASE)
            score += len(joins) * 10

            subqueries = re.findall(r"\(\s*SELECT\b", query.raw, re.IGNORECASE)
            score += len(subqueries) * 15

            aggregations = re.findall(r"\b(COUNT|SUM|AVG|MIN|MAX)\s*\(", query.raw, re.IGNORECASE)
            score += len(aggregations) * 5

        return score

    def _calculate_issue_complexity(self, issues: list[Issue]) -> int:
        """Calculate complexity based on detected issues."""
        score = 0

        for issue in issues:
            if issue.severity == Severity.CRITICAL:
                score += 25
            elif issue.severity == Severity.HIGH:
                score += 15
            elif issue.severity == Severity.MEDIUM:
                score += 10
            elif issue.severity == Severity.LOW:
                score += 5
            elif issue.severity == Severity.INFO:
                score += 2
        return score


class TrendTracker:
    """Tracks query complexity scores over time to calculate trends."""

    def __init__(self, cache_dir: Path | str = ".slowql_cache"):
        self.cache_dir = Path(cache_dir)
        self.trend_file = self.cache_dir / "complexity_trends.json"
        self._trends: dict[str, int] = {}
        self._load()

    def _load(self) -> None:
        """Load trends from file."""
        if self.trend_file.exists():
            try:
                self._trends = json.loads(self.trend_file.read_text(encoding="utf-8"))
            except Exception:
                self._trends = {}

    def _save(self) -> None:
        """Save trends to file."""
        try:
            self.cache_dir.mkdir(parents=True, exist_ok=True)
            self.trend_file.write_text(json.dumps(self._trends, indent=2), encoding="utf-8")
        except Exception:
            pass

    def get_trend(self, query_id: str, current_score: int) -> int | None:
        """
        Get trend for a query and update its stored score.

        Args:
            query_id: Unique identifier for the query (e.g., hash of normalized SQL).
            current_score: Current complexity score.

        Returns:
            Difference from last score (current - last), or None if no previous score.
        """
        last_score = self._trends.get(query_id)
        self._trends[query_id] = current_score
        self._save()

        if last_score is None:
            return None

        return current_score - last_score

File: slowql-rs/src/scoring.rs
use crate::models::issue::{Issue, Severity};
use crate::models::query::Query;

pub struct ComplexityScorer {
    base_score: u32,
}

impl ComplexityScorer {
    pub fn new() -> Self {
        ComplexityScorer { base_score: 10 }
    }

    pub fn calculate(&self, query: &Query, issues: &[Issue]) -> u32 {
        let mut score = self.base_score;

        // Structural complexity
        let upper = query.raw_upper();
        score += upper.matches("JOIN").count() as u32 * 10;
        score += upper.matches("(SELECT").count() as u32 * 15;
        score += upper.matches("COUNT(").count() as u32 * 5;
        score += upper.matches("SUM(").count() as u32 * 5;
        score += upper.matches("AVG(").count() as u32 * 5;

        // Issue complexity
        for issue in issues {
            score += match issue.severity {
                Severity::Critical => 25,
                Severity::High => 15,
                Severity::Medium => 10,
                Severity::Low => 5,
                Severity::Info => 2,
            };
        }

        score.min(100)
    }
}

impl Default for ComplexityScorer {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{Dimension, Location};

    fn make_query(sql: &str) -> Query {
        Query {
            raw: sql.to_string(),
            normalized: sql.to_string(),
            dialect: "postgresql".to_string(),
            location: Location::new(1, 1),
            ..Default::default()
        }
    }

    #[test]
    fn simple_query_low_score() {
        let scorer = ComplexityScorer::new();
        let q = make_query("SELECT id FROM users WHERE id = 1");
        let score = scorer.calculate(&q, &[]);
        assert!(score < 20);
    }

    #[test]
    fn complex_query_high_score() {
        let scorer = ComplexityScorer::new();
        let q = make_query("SELECT * FROM a JOIN b ON a.id=b.id JOIN c ON b.id=c.id JOIN d ON c.id=d.id WHERE EXISTS (SELECT 1 FROM e) AND x IN (SELECT id FROM f)");
        let score = scorer.calculate(&q, &[]);
        assert!(score > 40);
    }

    #[test]
    fn issues_increase_score() {
        let scorer = ComplexityScorer::new();
        let q = make_query("SELECT 1");
        let issues = vec![
            Issue::new("T-1", "test", Severity::Critical, Dimension::Security, Location::new(1,1), "x"),
            Issue::new("T-2", "test", Severity::High, Dimension::Security, Location::new(1,1), "x"),
        ];
        let score = scorer.calculate(&q, &issues);
        assert!(score >= 50);
    }

    #[test]
    fn score_capped_at_100() {
        let scorer = ComplexityScorer::new();
        let q = make_query("SELECT * FROM a JOIN b ON 1=1 JOIN c ON 1=1 JOIN d ON 1=1 JOIN e ON 1=1 JOIN f ON 1=1 JOIN g ON 1=1 JOIN h ON 1=1 JOIN i ON 1=1 JOIN j ON 1=1");
        let score = scorer.calculate(&q, &[]);
        assert_eq!(score, 100);
    }
}
